flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorSubtaskState.java

```diff
@@ -231,8 +232,7 @@ public void discardState() {
 			toDispose.addAll(rawOperatorState);
 			toDispose.addAll(managedKeyedState);
 			toDispose.addAll(rawKeyedState);
-			toDispose.addAll(inputChannelState);
-			toDispose.addAll(resultSubpartitionState);
+			toDispose.addAll(collectUniqueDelegates(inputChannelState, resultSubpartitionState));
```

Comment:
I think that's another place that would be better with FLINK-17972 - we could hide this logic inside the encapsulating `InFlightChannelStateHandle#discardState`

Reply:
Totally agree.